File: src/main/scala/Engine/Operators/Common/FlatMap/FlatMapTupleProcessor.scala
package Engine.Operators.Common.FlatMap

import Engine.Common.AmberTag.LayerTag
import Engine.Common.AmberTuple.Tuple
import Engine.Common.TupleProcessor

import scala.collection.Iterator

class FlatMapTupleProcessor(
    val flatMapFunc: (Tuple => TraversableOnce[Tuple]) with java.io.Serializable
) extends TupleProcessor {

  private var tupleIterator: Iterator[Tuple] = _
  private var nextFlag = false

  override def accept(tuple: Tuple): Unit = {
    this.tupleIterator = flatMapFunc.apply(tuple).toIterator
    this.nextFlag = tupleIterator.hasNext
  }

  override def onUpstreamChanged(from: LayerTag): Unit = {}

  override def onUpstreamExhausted(from: LayerTag): Unit = {}

  override def noMore(): Unit = {}

  override def initialize(): Unit = {}

  override def hasNext: Boolean = nextFlag

  override def next(): Tuple = {
    val next = tupleIterator.next
    this.nextFlag = tupleIterator.hasNext
    next
  }

  override def dispose(): Unit = {}
}

File: src/main/scala/Engine/Operators/Common/FlatMap/FlatMapTupleProcessor.java
package Engine.Operators.Common.FlatMap;

import Engine.Common.AmberTag.LayerTag;
import Engine.Common.AmberTuple.Tuple;
import Engine.Common.TupleProcessor;
import scala.Function1;
import scala.collection.Iterator;


public class FlatMapTupleProcessor implements TupleProcessor {

    private Iterator<Tuple> tupleIterator = null;
    private boolean nextFlag = false;
    private Function1<Tuple, Iterator<Tuple>> flatMapFunc;

    FlatMapTupleProcessor(Function1<Tuple, Iterator<Tuple>> flatMapFunc){
        this.flatMapFunc = flatMapFunc;
    }

    @Override
    public void accept(Tuple tuple) throws Exception {
        this.tupleIterator = flatMapFunc.apply(tuple);
        this.nextFlag = tupleIterator.toIterator().hasNext();
    }

    @Override
    public void onUpstreamChanged(LayerTag from) {
    }

    @Override
    public void onUpstreamExhausted(LayerTag from) {
    }

    @Override
    public void noMore() {
    }

    @Override
    public void initialize() throws Exception {
    }

    @Override
    public boolean hasNext() throws Exception {
        return nextFlag;
    }

    @Override
    public Tuple next() throws Exception {
        Tuple next = tupleIterator.next();
        this.nextFlag = tupleIterator.hasNext();
        return next;
    }

    @Override
    public void dispose() throws Exception {
    }
}
